src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+            {
+                casesBuilder.Add(previousCases[i]);
+            }
+
+            int index = patternIndex;
+            foreach ((BoundPattern pattern, SyntaxNode? syntax) in set.Cases)
+            {
+                var label = new GeneratedLabelSymbol("orCase");
+                SyntaxNode? diagSyntax = syntax;
+                if (diagSyntax is null)
+                {
+                    // TODO2 comment
+                    labelsToIgnore.Add(label);
+                    diagSyntax = nodeSyntax;
+                }
+
+                Debug.Assert(diagSyntax is not null);
+                casesBuilder.Add(builder.MakeTestsForPattern(++index, diagSyntax, rootIdentifier, pattern, whenClause: null, label: label));
+            }
+        }
+
+        /// <summary>
+        /// The purpose of this method is to bring `or` sequences to the top-level (so they can be used as separate cases).
+        /// Each set handles one `or`.
+        /// </summary>
+        private static SetsOfOrCases RewriteToSetsOfOrCases(BoundPattern? pattern)
+        {
+            return pattern switch
+            {
+                BoundBinaryPattern binary => rewriteBinary(binary),
+                BoundRecursivePattern recursive => rewriteRecursive(recursive),
+                BoundListPattern list => rewriteList(list),
+                BoundSlicePattern slice => rewriteSlice(slice),
+                BoundITuplePattern ituple => rewriteITuple(ituple),
+                BoundNegatedPattern => default,
+                BoundTypePattern => default,
+                BoundDeclarationPattern => default,
+                BoundConstantPattern => default,
+                BoundDiscardPattern => default,
+                BoundRelationalPattern => default,
+                null => default,
+                _ => throw ExceptionUtilities.UnexpectedValue(pattern)
+            };
+
+            static SetsOfOrCases rewriteBinary(BoundBinaryPattern binaryPattern)
+            {
+                SetsOfOrCases result = default;
+
+                if (binaryPattern.Disjunction)
+                {
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // In `A1 or ... or An`, we produce an expansion set: `A1`, ..., `An`
+                    OrCases resultOrSet1 = result.StartNewOrCases();
+                    foreach (var pattern in patterns)
+                    {
+                        resultOrSet1.Add(pattern, pattern.Syntax);
+                    }
+
+                    // If any of the nested patterns can be expanded, we carry those on.
+                    // For example, in `... or Ai or ...`, when we found multiple `or` patterns in `Ai` to be expanded
+                    // For each such nested expansion, we'll produce an `... or <expansion> or ...` expansion
+                    for (int i = 0; i < patterns.Count; i++)
```

Comment:
>for (int i = 0; i < patterns.Count; i++) [](http://example.com/codeflow?start=20&length=40)

It looks like this loop leads to unnecessary explosion of sets and to unnecessary complicated patterns in the cases of those sets. For example, for a pattern ```{ Prop1: not 42 or 43 } or { Prop2: not 44 or 45 }``` the result has three sets: 
```
Set:
{Prop1: ((not 42) or 43)}  => { Prop1: not 42 or 43 }
{Prop2: ((not 44) or 45)}  => { Prop2: not 44 or 45 }

Set:
({Prop1: (not 42)} or {Prop2: ((not 44) or 45)})  => 42
({Prop1: 43} or {Prop2: ((not 44) or 45)})  => 43

Set:
({Prop1: ((not 42) or 43)} or {Prop2: (not 44)})  => 44
({Prop1: ((not 42) or 43)} or {Prop2: 45})  => 45
```

However, I would expect the result to have one set:
```
Set:
{Prop1: (not 42)}  => 42
{Prop1: 43}  => 43
{Prop2: (not 44)}  => 44
{Prop2: 45}  => 45
```

The deeper the tree, the more overhead this is going to create.